File: docs-examples/example-java/src/test/java/io/micronaut/mqtt/docs/custom/type/ProductInfoSpec.java
package io.micronaut.mqtt.docs.custom.type;

import io.micronaut.mqtt.AbstractRabbitMQTest;
import io.micronaut.context.ApplicationContext;
import org.junit.jupiter.api.Test;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

public class ProductInfoSpec extends AbstractRabbitMQTest {

    @Test
    void testUsingACustomTypeBinder() {
        ApplicationContext applicationContext = startContext();

// tag::producer[]
        ProductClient productClient = applicationContext.getBean(ProductClient.class);
        productClient.send("body".getBytes(), new ProductInfo("small", 10L, true));
        productClient.send("body2".getBytes(), new ProductInfo("medium", 20L, true));
        productClient.send("body3".getBytes(), new ProductInfo(null, 30L, true));
// end::producer[]

        ProductListener productListener = applicationContext.getBean(ProductListener.class);

        try {
            await().atMost(5, SECONDS).until(() ->
                    productListener.messages.size() == 3 &&
                    productListener.messages.stream().anyMatch(pi -> pi.getCount() == 10L) &&
                    productListener.messages.stream().anyMatch(pi -> pi.getCount() == 20L) &&
                    productListener.messages.stream().anyMatch(pi -> pi.getCount() == 30L)
            );
        } finally {
            applicationContext.close();
        }
    }
}

File: test-suite-kotlin/src/test/kotlin/io/micronaut/mqtt/docs/custom/type/ProductInfoSpec.kt
package io.micronaut.mqtt.docs.custom.type

import io.micronaut.mqtt.AbstractMQTTTest
import org.awaitility.Awaitility.await
import org.junit.jupiter.api.Test
import java.util.concurrent.TimeUnit

class ProductInfoSpec : AbstractMQTTTest() {

    @Test
    fun testCustomTypeBinder() {
        val applicationContext = startContext()

        // tag::producer[]
        val productClient = applicationContext.getBean(ProductClient::class.java)
        productClient.send(ProductInfo("small", 10L, true))
        productClient.send(ProductInfo("medium", 20L, false))
        productClient.send(ProductInfo(null, 30L, true))
        // end::producer[]

        val productListener = applicationContext.getBean(ProductListener::class.java)

        try {
            await().atMost(5, TimeUnit.SECONDS).until {
                productListener.messages.size == 3 &&
                        productListener.messages.any { p -> p.size == "small" && p.count == 10L && p.sealed } &&
                        productListener.messages.any { p -> p.size == "medium" && p.count == 20L && !p.sealed } &&
                        productListener.messages.any { p -> p.size == null && p.count == 30L && p.sealed }
            }
        } finally {
            applicationContext.close()
        }
    }
}
